velox/exec/tests/utils/TestIndexStorageConnector.h

```diff
@@ -0,0 +1,299 @@
+    void initBuffer(BufferPtr& buffer, T*& rawBuffer, vector_size_t size) {
+      if (!buffer || !buffer->unique() ||
+          buffer->capacity() < sizeof(T) * size) {
+        buffer = allocateIndices(size, source_->pool_.get());
+      }
+      rawBuffer = buffer->asMutable<T>();
+    }
+
+    // Extracts the lookup result columns from the index table and return in
+    // 'result'.
+    void extractLookupColumns(
+        folly::Range<char* const*> rows,
+        RowVectorPtr& result);
+
+    // Inokved to trigger async lookup using background executor and return the
+    // 'future'.
+    void asyncLookup(vector_size_t size, ContinueFuture& future);
+
+    // Synchronously lookup the index table and return up to 'size' number of
+    // output rows in result.
+    std::unique_ptr<LookupResult> syncLookup(vector_size_t size);
+
+    const std::shared_ptr<TestIndexSource> source_;
+    const std::shared_ptr<TestIndexTable> indexTable_;
+    const LookupRequest request_;
+    const std::unique_ptr<HashLookup> lookupResult_;
+    folly::Executor* const executor_{nullptr};
+
+    std::atomic_bool hasPendingRequest_{false};
+    std::unique_ptr<BaseHashTable::JoinResultIterator> lookupResultIter_;
+    std::optional<std::unique_ptr<LookupResult>> asyncResult_;
+  };
+
+ private:
+  // Initialize the output projections for lookup result processing.
+  void initOutputProjections();
+
+  const std::shared_ptr<TestIndexTableHandle> tableHandle_;
+  const facebook::velox::RowTypePtr inputType_;
+  const facebook::velox::RowTypePtr outputType_;
+  const facebook::velox::RowTypePtr keyType_;
+  const facebook::velox::RowTypePtr valueType_;
+  facebook::velox::connector::ConnectorQueryCtx* const connectorQueryCtx_;
+  const std::shared_ptr<facebook::velox::memory::MemoryPool> pool_;
+  folly::Executor* const executor_;
+
+  std::vector<IdentityProjection> lookupOutputProjections_;
+
+  // The reusebale buffers for lookup result processing across lookup iterators.
+
+  // The input row number in lookup request for each matched result which is
+  // paired with 'outputRowMapping_' to indicate if a given input row has match
+  // or not. If the corresponding output row pointer in 'outputRowMapping_' is
+  // null, then there is no match for the given input row pointed by
+  // 'inputRowMapping_'.
+  BufferPtr inputRowMapping_;
+  vector_size_t* rawInputRowMapping_{nullptr};
+  // Points to the matched row pointer in 'indexTable_' for each input row. If
+  // there is a miss for a given input, then this is set to null.
+  BufferPtr outputRowMapping_;
+  char** rawOutputRowMapping_{nullptr};
+  // The input row number in request for each output row in the returned lookup
+  // result. Any gap in the input row numbers means the corresponding input rows
+  // that has no matches in the index table.
+  BufferPtr inputHitIndices_;
+  vector_size_t* rawInputHitIndices_{nullptr};
+
+  RowVectorPtr lookupOutput_;
+  friend class ResultIterator;
```

Comment:
Let's not use 'friend'. It couples tests with implementation details too much and makes it difficult to refactor the code.